Add passes and plugins for distributed inference of NLU

wanghaoshuang opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
New features 

### PR changes
Others

### Describe
Add some passes and  plugins for distributed inference
1. Add a pass to delete c_identity_op
2. Add a pass to fuse elementwise+elementwise+layernorm to preln_residual_bias op
3. Add c_allreduce trt plugin and preln_residual_bias trt plugin

To be fixed: test_trt_convert_preln_residual_bias.py

qingqing01 committed:
Support MP based TensorRT

qingqing01 committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into develop

wanghaoshuang committed:
Merge branch 'tensor_nccl' of https://github.com/qingqing01/Paddle into dist_trt

wanghaoshuang committed:
Add pass to remove c_identity op and fix some issues of c_allreduce_op_plugin

wanghaoshuang committed:
Fix ele add

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into dist_trt

wanghaoshuang committed:
Add preln_residual_bias_fuse_pass to fuse pre-layernorm block

wanghaoshuang committed:
Add preln_residual_bias_plugin

wanghaoshuang committed:
Fix preln_residual_bias

wanghaoshuang committed:
hard coding for test int8 inference of BERT

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into dist_trt_nlu

wanghaoshuang committed:
Revert some modifies

wanghaoshuang committed:
Add unittest

qingqing01 commented on paddle/fluid/inference/tensorrt/convert/c_allreduce_op.cc:
```diff
@@ -0,0 +1,103 @@
+    int input_num = op_desc.Input("X").size();
+    PADDLE_ENFORCE_EQ(input_num, 1,
+                      platform::errors::InvalidArgument(
+                          "The input X's size must equal to 1 in TRT swish op."
```
in TRT swish op  -> in  c_allreduce_sum op.

wanghaoshuang replied:
Done.

qingqing01 commented on paddle/fluid/inference/tensorrt/convert/c_allreduce_op.cc:
```diff
@@ -0,0 +1,103 @@
+    PADDLE_ENFORCE_EQ(
+        output_num, 1UL,
+        platform::errors::InvalidArgument(
+            "The ouput Out's size must equal to 1 in TRT swish op. "
```
同上

qingqing01 commented on paddle/fluid/inference/tensorrt/convert/preln_residual_bias.cc:
```diff
@@ -0,0 +1,98 @@
+                  const framework::Scope& scope, bool test_mode) override {
+    VLOG(4) << "convert fused preln_residual_bias op to tensorrt layer";
+    std::cout << "convert fused preln_residual_bias op to tensorrt layer"
+              << std::endl;
```
remove

wanghaoshuang replied:
Removed.

qingqing01 commented on paddle/fluid/inference/tensorrt/convert/preln_residual_bias.cc:
```diff
@@ -0,0 +1,98 @@
+    auto input2_dims = input2->getDimensions();
+    for (int i = 0; i < input2_dims.nbDims; i++) {
+      std::cout << "input2 dims: " << input2_dims.d[i] << std::endl;
+    }
```
remove debug code

wanghaoshuang replied:
Removed.

qingqing01 commented on paddle/fluid/inference/tensorrt/convert/c_allreduce_op.cc:
```diff
@@ -0,0 +1,103 @@
+                  const framework::Scope& scope, bool test_mode) override {
+    VLOG(4) << "convert fluid callreduce op to tensorrt layer";
+    ReduceType red_type = op_to_reduce_type[op.type()];
+    std::string name = "c_allreduce_sum";
```
这里name注意下是否正确？ 这个Converter是个sum/min/max/prod公用的

qingqing01 commented on paddle/fluid/inference/tensorrt/plugin/c_allreduce_op_plugin.cu:
```diff
@@ -0,0 +1,240 @@
+  if (for_test_) {
+    VLOG(4) << "skip communication for unittest.";
+    return 0;
+  }
```
这个是debug code？

wanghaoshuang replied:
不是debug code, 这里的`for_test_`用于简化单测。
如果不跳过通信操作，则需要构造一个很复杂的单测，暂时还没找到合适的方法。

wanghaoshuang replied:
已删除，并补充单测。

qingqing01 commented on python/paddle/fluid/tests/unittests/ir/inference/test_trt_convert_c_allreduce.py:
```diff
@@ -0,0 +1,96 @@
+                                                                     True), 1e-5
+
+    def test(self):
+        self.run_test(skip_baseline=True)
```
add comments why skip this

wanghaoshuang replied:
Done.

qingqing01 commented on python/paddle/fluid/tests/unittests/ir/test_ir_preln_residual_bias_fuse_pass.py:
```diff
@@ -0,0 +1,57 @@
+                name="y", shape=[128, 768], dtype="float32", lod_level=0)
+            x = fluid.layers.elementwise_add(x=x, y=bias)
+            elementwise_out = fluid.layers.elementwise_add(x=x, y=y)
+            out = fluid.layers.layer_norm(input=elementwise_out)
```
use 2.0 API

wanghaoshuang replied:
Fixed.

qingqing01 commented on paddle/fluid/inference/api/paddle_pass_builder.cc:
```diff
@@ -96,26 +96,25 @@ const std::vector<std::string> kTRTSubgraphPasses({
+      "delete_c_identity_op_pass",                    //
       "multihead_matmul_fuse_pass_v2",                //
       "multihead_matmul_fuse_pass_v3",                //
+      "preln_residual_bias_fuse_pass",                //
```
对于非多卡大模型，如果先匹配这个pass，是否影响原先的推理，这块确认下~

wanghaoshuang replied:
将该pass移到了skip_layernorm相关pass之后

qingqing01 commented on paddle/fluid/inference/api/paddle_pass_builder.cc:
```diff
@@ -96,26 +96,25 @@ const std::vector<std::string> kTRTSubgraphPasses({
       "simplify_with_basic_ops_pass",                 //
       "embedding_eltwise_layernorm_fuse_pass",        //
       "preln_embedding_eltwise_layernorm_fuse_pass",  //
+      "trt_map_matmul_v2_to_mul_pass",                //
```
和下面重复了

wanghaoshuang committed:
Fix format

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into dist_trt_nlu

wanghaoshuang committed:
Fix unitest

wanghaoshuang committed:
Fix unittest

wanghaoshuang committed:
Fix code style

wanghaoshuang committed:
check PADDLE_WITH_NCCL in c_allreduce_op_plugin

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into dist_trt_nlu

qingqing01 commented on python/paddle/fluid/tests/unittests/ir/inference/test_trt_c_allreduce_infer_script.py:
```diff
@@ -0,0 +1,108 @@
+                        })
+        out = paddle.static.nn.fc(x=c_data, size=1)
+        mean = paddle.mean(out)
+    print(main_program)
```
remove

qingqing01 commented on python/paddle/fluid/tests/unittests/ir/inference/test_trt_c_allreduce_infer_script.py:
```diff
@@ -0,0 +1,108 @@
+        output_handle = predictor.get_output_handle(output_names[0])
+        output_data = output_handle.copy_to_cpu()  # numpy.ndarray类型
+        print("Output data size is {}".format(output_data.size))
+        print("Output data is {}".format(output_data))
```
可以check 输出的shape，以及输出的正确性吧

qingqing01 commented on paddle/fluid/inference/tensorrt/plugin/c_allreduce_op_plugin.cu:
```diff
@@ -0,0 +1,239 @@
+  PADDLE_ENFORCE_GPU_SUCCESS(platform::dynload::ncclAllReduce(
+      sendbuff, recvbuff, numel, dtype, nccl_red_type, comm->comm(), stream));
+#endif
+  return (cudaGetLastError() != cudaSuccess);
```
对于报错这块，如果inference编译不包含NCCL，需要明确的提示下~

qingqing01 commented on paddle/fluid/inference/tensorrt/plugin/c_allreduce_op_plugin.cu:
```diff
@@ -0,0 +1,239 @@
+// int CAllReducePluginDynamic::initialize() TRT_NOEXCEPT {
+//  getPluginNamespace();
+//  return 0;
+//}
```
这块如果没用就删掉

qingqing01 commented on python/paddle/fluid/tests/unittests/ir/inference/test_trt_convert_skip_layernorm.py:
```diff
@@ -190,15 +190,15 @@ def generate_trt_nodes_num(attrs, dynamic_shape):
+        #     attrs, False), 1e-5
+        # self.trt_param.precision = paddle_infer.PrecisionType.Half
+        # yield self.create_inference_config(), generate_trt_nodes_num(
+        #     attrs, False), 1e-5
```
why remove these code?

minghaoBD commented on paddle/fluid/inference/tensorrt/convert/preln_residual_bias.cc:
```diff
@@ -0,0 +1,90 @@
+
+    auto input1_dims = input1->getDimensions();
+    for (int i = 0; i < input1_dims.nbDims; i++) {
+      std::cout << "input1 dims: " << input1_dims.d[i] << std::endl;
```
这个debug信息可以删除下？或者用VLOG(4)

minghaoBD commented on paddle/fluid/inference/tensorrt/plugin/c_allreduce_op_plugin.cu:
```diff
@@ -0,0 +1,239 @@
+  if (input_type == nvinfer1::DataType::kFLOAT) {
+    VLOG(1) << "TRT Plugin DataType selected. CAllReduce-->fp32";
+    sendbuff = const_cast<void*>(inputs[0]);
+  } else if (input_type == nvinfer1::DataType::kHALF) {
```
这里不用考虑int8的情况吗？

wanghaoshuang replied:
暂不支持Int8

wanghaoshuang committed:
Fix some issues:
1. Fix device memory access illegal error in preln_residual_bias_plugin
2. Make preln_residual_bias_plugin support FP32
3. Skip some unittests when WITH_DISTRIBUTE is OFF
4. Check result in unittests of c_allreduce_plugin
5. Remove unused lines
6. Add warning for nccl in c_allreduce_plugin

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into dist_trt_nlu

wanghaoshuang committed:
Fix unittest and code format

wanghaoshuang committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/Paddle into dist_trt_nlu

wanghaoshuang committed:
Fix code coverage

XiaoguangHu01 reviewed (approved):
LGTM
